ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c

```diff
@@ -6205,3 +6219,441 @@ bdb_back_ctrl(Slapi_Backend *be, int cmd, void *info)
+     *  exiting the function (no more need to free the
+     *  value if its address change as it is the case with 
+     *  when using DB_DBT_MALLOC )
+     */
+    if (dbt->data != dbi->data) {
+        if (dbi->flags & DBI_VF_DONTGROW) {
+            /* trying to realloc a user buffer */
+            PR_ASSERT(0);
+            dblayer_value_set_buffer(bdb_be(), dbi, (void*)(-1), -1);
+            return;
+        }
+        if (dbt->flags & (DB_DBT_MALLOC | DB_DBT_REALLOC)) {
+            if (isresponse) {
+                dbi->data = NULL; /* Value is already freed by dbt realloc */
+            }
+            dblayer_value_set(bdb_be(), dbi, dbt->data, dbt->size);
+            dbt->data = NULL;  /* Insure that value will not be freed through dbt */
+        } else if (dbt->flags & DB_DBT_USERMEM) {
+            dblayer_value_set_buffer(bdb_be(), dbi, dbt->data, dbt->size);
+            dbi->ulen = dbt->ulen;
+        } else {
+            /* trying to use uninitialized DBT */
+            PR_ASSERT(0);
+            dblayer_value_set_buffer(bdb_be(), dbi, (void*)(-1), -1);
+            return;
+        }
+        dbi->ulen = dbt->ulen;
+    } else {
+        /* data buffer has not changed ==> update the size */
+        dbi->size = dbt->size;
+        dbi->ulen = dbt->ulen;
+    }
+}
+
+/**********************/
+/* dbimpl.c callbacks */
+/**********************/
+
+char *bdb_public_get_db_id(dbi_db_t *env)
+{
+    return ((DB*)env)->fname;
+}
+
+int bdb_public_bulk_free(dbi_bulk_t *bulkdata)
+{
+    /* No specific action required for berkeley db handling */
+    return DBI_RC_SUCCESS;
+}
+
+int bdb_public_bulk_nextdata(dbi_bulk_t *bulkdata, dbi_val_t *data)
+{
+    DBT bulk;
+    void *retdata = NULL;
+    u_int32_t retdlen = 0;;
+    bdb_dbival2dbt(&bulkdata->v, &bulk, PR_FALSE);
+    if (bulkdata->v.flags & DBI_VF_BULK_DATA) {
+        DB_MULTIPLE_NEXT(bulkdata->it, &bulk, retdata, retdlen);
+        dblayer_value_set_buffer(bulkdata->be, data, retdata, retdlen);
+    } else {
+        /* Coding error - bulkdata is not initialized or wrong type */
+        PR_ASSERT(0);
+        return DBI_RC_UNSUPPORTED;
+    }
+    if (retdata == NULL || bulkdata->be == NULL) {
+        return DBI_RC_NOTFOUND;
+    }
+    return DBI_RC_SUCCESS;
+}
+
+int bdb_public_bulk_nextrecord(dbi_bulk_t *bulkdata, dbi_val_t *key, dbi_val_t *data)
+{
+    DBT bulk;
+    void *retkey = NULL;
+    void *retdata = NULL;
+    u_int32_t retklen = 0;;
+    u_int32_t retdlen = 0;;
+    bdb_dbival2dbt(&bulkdata->v, &bulk, PR_FALSE);
+    if (bulkdata->v.flags & DBI_VF_BULK_RECORD) {
+        DB_MULTIPLE_KEY_NEXT(bulkdata->it, &bulk, retkey, retklen, retdata, retdlen);
+        dblayer_value_set_buffer(bulkdata->be, data, retdata, retdlen);
+        dblayer_value_set_buffer(bulkdata->be, key, retkey, retklen);
+    } else {
+        /* Coding error - bulkdata is not initialized or wrong type */
+        PR_ASSERT(0);
+        return DBI_RC_UNSUPPORTED;
+    }
```

Comment:
No need to call bdb_dbt2dbival (bulk->bulkdata) ?

Reply:
same as above